holoviews/operation/datashader.py

```diff
@@ -58,13 +59,19 @@ class AggregationOperation(ResampleOperation2D):
     aggregator parameter used to define a datashader Reduction.
     """
 
-    aggregator = param.ClassSelector(class_=(ds.reductions.Reduction, str),
-                                     default=ds.count(), doc="""
+    aggregator = param.ClassSelector(class_=(rd.Reduction, rd.summary, str),
+                                     default=rd.count(), doc="""
         Datashader reduction function used for aggregating the data.
         The aggregator may also define a column to aggregate; if
         no column is defined the first value dimension of the element
         will be used. May also be defined as a string.""")
 
+    selector = param.ClassSelector(class_=(rd.min, rd.max, rd.first, rd.last),
```

Comment:
Assume I'm stupid, can you explain to me how aggregator, selector and where as an aggregator interact?

Reply:
A selector is a subset of the aggregator function, which selects an item for the data / dataframe. With this PR, we support `rd.first`, `rd.last`, `rd.min`, and `rd.max` in HoloViews. For the next examples, we have a dataframe with two columns `s` and `val`, which are both numeric.

`rd.where` can use a selector to look up an index. `rd.where(rd.min("val"))` will find the index of the minimum value. By passing in a column like `rd.where(rd.min("val"), "s")` will find the s-value where the `val` has its minimum value.

The selectors can be passed into the rasterize-aggregator parameter with or without the where. Suppose we want to use an aggregator which is not a selector like `rd.Count` but still wants some information about the underlying data. We will use the rasterize-selector parameter and pass in `ds.first("val")`. Behind the scene, this will then use `rd.summary` to combine the aggregator `rd.Count` with `ds.where(ds.first("val"))` and only do one pass one data shader pass on data and return the output of both.

A special thing is done in HoloViews when passing in `rd.where(rd.first("val")` into rasterize, where it will not return the index itself but all of the row data for that index. This means in the previous example, we will get a hover tool containing `Count` (from the aggregator), `val`, and `s` (from the selector).

Reply:
Thank you, **I think** that all makes sense to me but we will have to have detailed examples to explain this.

Reply:
@Hoxbro can correct me if I am wrong, but here is my simplified understanding of how to think about this:

The selector is used to specify which index value is used to build all the `Dataset` layers (e.g `first` or `max` on some dimension) and this needs to be combined with the `aggregator` (which may or may not be a selector) in `summary` to do a single pass. That way the aggregator (e.g. `mean`) which is not a selector can be put on the bottom layer (to make it visible and plotted) in the `Dataset` and the rest of the layers (computed from the selector) comes from the supplied selector.

What `where` does is make something like `max('foo')` go from giving you the max value of `'foo'` to the max *index* of foo. This is done automatically and internally when supplying `max('foo')` as the selector.

Hope that makes sense!

What I don't yet get is what happens if you *explicitly* supply a `where` aggregator as a selector. I think this is what @Hoxbro is talking about with '... where it will not return the index itself but all of the row data for that index.' but I have to admit I don't really understand what that means yet...

Reply:
> A special thing is done in HoloViews when passing in rd.where(rd.first("val") into rasterize, where it will not return the index itself but all of the row data for that index. 

I think this needs to be configurable; returning all the row data provides good pixel-to-pixel interactivity but if there are many rows or some rows are structured data like polygons or text the result may slow down the aggregation step significantly. So I think we need a good default here that someone can override when they disagree with the default. 

My gut tells me that returning the index and then dynamically requesting the row will give the best user experience, in that it makes all the information available without any special setup and without slowing down aggregations. But I don't think that's what it does now (and would require a custom hover tool?), and also won't always be optimal, since when there are only a couple of hover fields and they are simple scalars (e.g. categories) then it makes sense to send that data to the client as the complete screen-sized array. Plus then if it's not a dynamic plot hover will still be fully supported. 

So, I think this behavior should be configurable, and after it's configurable we can decide on the default, which my _guess_ (without actually trying it) would be the index array rather than all the row's columns.